### PR TITLE
Fix computation in normalization layers

### PR DESCRIPTION
**Pull request by AakashKumarNain**
Irrespective of the dtype passed to the normalization layers, the calculations should be done with fp32 or higher dtype. Not doing that can cause instabilities in training which is too common when someone is training models in half-precision. Now, we can delegate this to the end user to do this explicitly, but it is an easy thing to miss. This PR takes care of that.

### CODE REVIEW

**Comment by patrick-kidger** [comment]
Looks like the current implementation doesn't pass under strict-dtype-promotion (see failing tests), but other than that this sounds good to me!

**Comment by AakashKumarNain** [comment]
Thanks @patrick-kidger I have fixed it. Please let me know if you have any more suggestions on this one

File: equinox/nn/_normalisation.py [hunk 1]
```diff
@@ -94,6 +94,10 @@ def __init__(
             )
         self.use_weight = use_weight
         self.use_bias = use_bias
+
+        with jax.numpy_dtype_promotion("standard"):
+            dtype = jnp.result_type(dtype, jnp.float32)
```
Comment (patrick-kidger) [inline equinox/nn/_normalisation.py:99]: This one seems wrong to me. It's totally valid for a user to want to set the weight and bias to a lower dtype, if they really want to.

Comment (AakashKumarNain) [reply equinox/nn/_normalisation.py:99]: Makes sense. Changed it.

**Comment by patrick-kidger** [comment]
Alright, this LGTM! Final question before I merge this, do you have a reference for needing higher precision for these operations? (As much for me to read up on as anything else :D )

**Comment by AakashKumarNain** [comment]
I think this is pretty much enough: https://pytorch.org/docs/stable/amp.html#autocast-op-reference

**Comment by patrick-kidger** [comment]
Alright, looks good to me! Thanks for spotting this, and for the reference. Merged :)
